core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -20,12 +20,44 @@
 package org.elasticsearch.common.logging;
 
 import org.elasticsearch.common.SuppressLoggerChecks;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
 
 /**
  * A logger that logs deprecation notices.
  */
 public class DeprecationLogger {
 
+    /**
+     * The "Warning" Header comes from RFC-7234. As the RFC describes, it's generally used for caching purposes, but it can be
+     * used for <em>any</em> warning.
+     *
+     * https://tools.ietf.org/html/rfc7234#section-5.5
+     */
+    public static final String DEPRECATION_HEADER = "Warning";
+
+    /**
+     * This is set once by the {@code Node} constructor, but it uses {@link CopyOnWriteArraySet} to ensure that
+     * tests can run in parallel.
+     */
+    private static final CopyOnWriteArraySet<ThreadContext> THREAD_CONTEXT = new CopyOnWriteArraySet<>();
+
+    /**
+     * Set the {@link ThreadContext} used to add deprecation headers to network responses.
+     * <p>
+     * This is expected to <em>only</em> be invoked by the {@code Node}'s constructor (therefore once outside of tests).
+     *
+     * @param threadContext The thread context owned by the {@code ThreadPool} (and implicitly a {@code Node})
+     */
+    public static void setThreadContext(ThreadContext threadContext) {
+        assert threadContext != null;
+
+        THREAD_CONTEXT.add(threadContext);
```

Comment:
In case this is left as a set, can we add a check to throw an exception if the set already contains the ThreadContext